tests/baselines/reference/ramdaToolsNoInfinite.errors.txt

```diff
@@ -0,0 +1,126 @@
+            ((head: E, ...args: T) => any) extends ((...args: infer U) => any)
+            ? U
+            : T;
+    
+        type Drop<N extends number, T extends any[], I extends any[] = []> = {
+            0: Drop<N, Tail<T>, Prepend<any, I>>;
+            1: T;
+        }[
+            Length<I> extends N
+            ? 1
+            : 0
+        ];
+    
+        type Cast<X, Y> = X extends Y ? X : Y;
+    
+        type Pos<I extends any[]> =
+            Length<I>;
+    
+        type Next<I extends any[]> =
+            Prepend<any, I>;
+    
+        type Prev<I extends any[]> =
+            Tail<I>;
+    
+        type Iterator<Index extends number = 0, From extends any[] = [], I extends any[] = []> = {
+            0: Iterator<Index, Next<From>, Next<I>>;
+            1: From;
+        }[
+            Pos<I> extends Index
+            ? 1
+            : 0
+        ];
+    
+        type Reverse<T extends any[], R extends any[] = [], I extends any[] = []> = {
+            0: Reverse<T, Prepend<T[Pos<I>], R>, Next<I>>;
+            1: R;
+        }[
+            Pos<I> extends Length<T>
+            ? 1
+            : 0
+        ];
+    
+        type Concat<T1 extends any[], T2 extends any[]> =
+            Reverse<Reverse<T1> extends infer R ? Cast<R, any[]> : never, T2>;
+    
+        type Append<E, T extends any[]> =
+            Concat<T, [E]>;
+    
+        type ValueOfRecord<R> = R extends Record<any, infer T> ? T : never;
+    }
+    
+    declare namespace R {
+        export type Placeholder = { __placeholder: void };
+    }
+    
+    declare namespace Curry {
+        type GapOf<T1 extends any[], T2 extends any[], TN extends any[], I extends any[]> =
+            T1[Tools.Pos<I>] extends R.Placeholder
+            ? Tools.Append<T2[Tools.Pos<I>], TN>
+            : TN;
+    
+        interface GapsOfWorker<T1 extends any[], T2 extends any[], TN extends any[] = [], I extends any[] = []> {
+            0: GapsOf<T1, T2, GapOf<T1, T2, TN, I> extends infer G ? Tools.Cast<G, any[]> : never, Tools.Next<I>>;
+            1: Tools.Concat<TN, Tools.Drop<Tools.Pos<I>, T2> extends infer D ? Tools.Cast<D, any[]> : never>;
+        }
+        type GapsOf<T1 extends any[], T2 extends any[], TN extends any[] = [], I extends any[] = []> = GapsOfWorker<T1, T2, TN, I>[
+            Tools.Pos<I> extends Tools.Length<T1>
+            ? 1
+            : 0
+        ];
+    
+        type PartialGaps<T extends any[]> = {
+            [K in keyof T]?: T[K] | R.Placeholder
+        };
+    
+        type CleanedGaps<T extends any[]> = {
+            [K in keyof T]: NonNullable<T[K]>
+        };
+    
+        type Gaps<T extends any[]> = CleanedGaps<PartialGaps<T>>;
+    
+        type Curry<F extends ((...args: any) => any)> =
+            <T extends any[]>(...args: Tools.Cast<Tools.Cast<T, Gaps<Parameters<F>>>, any[]>) =>
+                              ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

```

Comment:
How is the change in the PR causing this error?

Reply:
I believe the underlying issue is the same as it was in my PRs. This is now resolved through layers and thus `getResolvedApparentTypeOfMappedType` sees the constraint of `Parameters<F>` - that once gets computed as `unknown[]` (by `getDefaultConstraintOfConditionalType` when restrictive instantiation bails out from `getConstraintOfDistributiveConditionalType`) and once as `any` (by `getConstraintOfDistributiveConditionalType`). So `any` can get returned and `getResolvedApparentTypeOfMappedType` can't recognize this as an array-like. So it ends up not d about it not being assignable to `readonly any[]` (that is being checked to determine if the type can be used as rest). 

IIRC the other case (when `unknown[]`) is returned makes `any[]` (added by this `Cast`) eliminated from the substitution type since, at that point, it's determined that the `baseType` satisfies the added constraint - so the `baseType` gets returned.

Reply:
So the issue is mainly in the fact that previously the substitution type wasn't simplified at all here and that `& any[]` in it (coming from its added constraint) was preserved and made the comparison against `readonly any[]` OK.

Reply:
What do you mean by "this is now resolved through layers"?
How do we go from a call to `getResolvedApparentTypeOfMappedType` to calls to `getDefaultConstraintOfConditionalType` and `getConstraintOfDistributiveConditionalType`?

Reply:
Please note that I'm partially speaking based on what I remember about debugging this last week - so some details might be slightly off.

`Gaps` is a mapped type - so we enter `getResolvedApparentTypeOfMappedType`. With this PR we recognize that an instantiation of a homomorphic mapped type is received and we end up checking its constraint and if it's an index type. It turns out it is because it's something like `keyof Parameters<...>`. So we grab the `.type` and compute its base constraint. So now we are computing the constraint of a conditional type (through `getConstraintFromConditionalType`) and, as mentioned above, this once resolves through `getDefaultConstraintOfConditionalType` (for the restrictive instantiation case) and once through `getConstraintOfDistributiveConditionalType` (for a regular case :P).

Reply:
With latest commit this error is gone.